### ISSUE TITLE
torch.autograd.grad output variable has no grad_fn

### ISSUE THREAD

**Issue by seerdecker** [pytorch/pytorch#5479]
See the code below. I would expect d.grad_fn not to be None.
```
a = torch.autograd.Variable(torch.FloatTensor([[1, 2, 3], [4, 5, 6]]), requires_grad=1)
b = torch.mul(a, 2)
c = b.sum()
d = torch.autograd.grad(c, b, torch.FloatTensor([2]), create_graph=True)[0]

print("d gradients are fine")
print(d)

print("d.grad_fn is None")
print(d.grad_fn)

print("Boom")
d.sum().backward()
```

Testing on Linux, Pytorch 0.3.1, Python 3.6.

**Comment by colesbury** [comment]
`dC/dB` is constant, so there's no second derivative

**Comment by apaszke** [comment]
`d` is not even a leaf variable (not created manually using the constructor), so it won't get grad saved by default. If you want to get it call `.retain_grad()`

**Comment by seerdecker** [comment]
`dC/dB` exists but is zero.

Note that this code doesn't throw even though the gradient is zero
`(torch.autograd.Variable(torch.FloatTensor([[1, 2, 3], [4, 5, 6]]), requires_grad=1)*0).sum().backward()`

I'll argue that the current semantics are not very helpful to the user when debugging. If you change the loss and cause the gradient to become zero, the code throws. It's kind of like having to write `if (ptr) free(ptr)` instead of `free(ptr)`. Not a big deal, but annoying.
